app/controllers/tcsp/SummaryController.scala

```diff
@@ -37,11 +38,28 @@ class SummaryController @Inject()
   val statusService: StatusService
 ) extends BaseController {
 
+  def sortProviders(data: Tcsp): List[String] = {
+
+    val res = data.tcspTypes match {
+      case Some(types) => types.serviceProviders.collect{
+        case provider if provider.value != "05" => Messages(s"tcsp.service.provider.lbl.${provider.value}")
+      }
+    }
+
+    val special = (data.tcspTypes match {
+      case Some(types) => types.serviceProviders.filter { case provider => provider.value == "05" }
+    }) map {
+      case _ => Messages("tcsp.service.provider.lbl.05")
+    }
+
+    res.toList.sorted ++ special.toList
+  }
+
```

Comment:
what if tcspTypes is empty? data.tcspTypes matching will fail. Will you pass then empty list? Will the view display the question with empty answer?

Reply:
this should never happen as the frontend wont allow the user to select nothing.
we can be defensive in out coding, but if this turn out to cause a problem, then the issue is with the ui page validators not the responsibility of this function

Reply:
at the moment it is possible to get into situation where summary page will try to display the questions and answers even when there's no answers in database, it's possible using back button:
1. add tcsp and answer all questions
2. remove tcsp
3. go back using back button to check your answers
you will see questions with empty answers

this is problem with check your answers itself which should not be displayed if section is incomplete, but proves that our frontend is not bulletproof

Reply:
so we could leave it as it is, because we have ticket in backlog to address this issue, but as long as it's not solved I don't think we should leave this option to display empty answers

Reply:
the back-link problem does not happen with latest code merged from master

Reply:
I understand the effect this can have if the list if empty.
coming from the mindset of **single responsibility principle**, I would say something else should handle the fact that the list is empty to start with. hence, the sortedList function should not get called. 
However, in this case we have few protection layers,

- the frontend wont allow you to leave the page without selecting at least one.
- amls wont allow you submit as the section will be incomplete.

hope that answers your concerns

Reply:
sortedList function above can return List of strings or empty list... interpreting the signature of the function it should return sorted list... what is then the default order of empty list ascending or descending? if you cannot answer it means this function is doing two things: returning sorted list or empty list, so there are two possible results, so single responsibility is violated here as well :) should this be handled somehow before calling the sortedList function... maybe but it's not. Unfortunately model is constructed so that the TcspTypes are optional, so empty list is possible and valid outcome and the schema accepts that happily...

Reply:
@ajchilts ^^

Reply:
Ultimately test the solution. If we identify a scenario that breaks the code we should either; 

i). Fix the scenario so it is not possible to break the code or;
ii). Add some defensive coding to handle it. 
iii). Both!

Despite what is stated in various standards I have seen too many instances where we get caught out by a scenario that 'should not exist'. With this in mind I do tend to code defensively. Remember; the front end is not the only manipulator of data! We have ETMP in play and we have been stung by schema mismatches and missing mongo data. I do not have a problem with devs coding defensively as 500 'please try again later' messages are a bad user experience. 

While standards are here for a reason and very important please put the user journey at the forefront of you mind when coding solutions. Of course also adhering to the standards where possible.

Reply:
Following andys comment we should then revert to swavs commit and pass this ticket on since it covers all cases? In my opinion to much time has been spent on this ticket already. The functionality worked in the first place, passed tests and was secure. We've spent an extra day working on this.

Reply:
a quick mobbing session with everyone, we decided to keep the unit tested code and log an error if the list is empty